### PR TITLE
DIV-5581 extend idam client / use wiremock JRE

### PR DESCRIPTION
**Pull request by alexwakeman**
### JIRA link (if applicable) ###
DIV-5581 [Use idam-java-client lib instead of our own idam-client](https://tools.hmcts.net/jira/browse/DIV-5581)

### Change description ###
- Use WireMock JRE: allow for stub configuration in code, easier to add tests, no dependencies on server, can run in pipeline
- Extend API to cate for Divorce needs
- No breaking changes

**Does this PR introduce a breaking change?**
```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/idam/client/IdamApi.java [hunk 1]
```diff
@@ -17,15 +19,35 @@
...
-    @RequestMapping(method = RequestMethod.POST, value = "/pin")
+    @RequestMapping(
+        method = RequestMethod.GET,
+        value = "/details"
+    )
+    UserDetails retrieveUserDetails(
+        @RequestHeader(HttpHeaders.AUTHORIZATION) String authorisation
+    );
+
+    @RequestMapping(
```
Comment (timja) [inline src/main/java/uk/gov/hmcts/reform/idam/client/IdamApi.java]: PostMapping would be simpler if you're going to touch this code,

and GetMapping on the rest on the Gets..

File: src/main/java/uk/gov/hmcts/reform/idam/client/models/UserDetails.java [hunk 2]
```diff
@@ -3,22 +3,27 @@
...
+    private String email;
+    private String forename;
+    private String surname;
+    private List<String> roles;
 
-    private final String forename;
-    private final String surname;
-    private final List<String> roles;
+    public UserDetails() {
+        super();
```
Comment (timja) [inline src/main/java/uk/gov/hmcts/reform/idam/client/models/UserDetails.java]: unneeded?
```suggestion
```

Comment (alexwakeman) [reply src/main/java/uk/gov/hmcts/reform/idam/client/models/UserDetails.java]: this is needed for WireMock stubs to work

File: src/main/java/uk/gov/hmcts/reform/idam/client/CoreFeignConfiguration.java [hunk 3]
```diff
@@ -8,12 +8,9 @@
 import org.springframework.boot.autoconfigure.http.HttpMessageConverters;
 import org.springframework.cloud.openfeign.support.SpringEncoder;
 import org.springframework.context.annotation.Bean;
-import org.springframework.context.annotation.Configuration;
 import org.springframework.context.annotation.Primary;
 import org.springframework.context.annotation.Scope;
 
-
-@Configuration
```
Comment (hosainnet) [inline src/main/java/uk/gov/hmcts/reform/idam/client/CoreFeignConfiguration.java]: Is this expected?

Comment (alexwakeman) [reply src/main/java/uk/gov/hmcts/reform/idam/client/CoreFeignConfiguration.java]: Yes - we want to inherit the Configuration of the host project, not have our own here, so we can apply the config from there.

File: config/checkstyle/checkstyle.xml [hunk 4]
```diff
@@ -179,7 +180,7 @@
     </module>
     <module name="AbbreviationAsWordInName">
       <property name="ignoreFinal" value="false"/>
-      <property name="allowedAbbreviations" value="JNDI,UUID"/>
+      <property name="allowedAbbreviationLength" value="40"/>
```
Comment (alexwakeman) [inline config/checkstyle/checkstyle.xml]: Using CAPS_PROPS in the Test file - better for readability.

File: src/main/java/uk/gov/hmcts/reform/idam/client/CoreFeignConfiguration.java (hunk 3 above)
Comment (alexwakeman) [reply src/main/java/uk/gov/hmcts/reform/idam/client/CoreFeignConfiguration.java]: Also, this is Feign config, so would not apply to Spring...

File: config/checkstyle/checkstyle.xml (hunk 4 above)
Comment (timja) [reply config/checkstyle/checkstyle.xml]: What does this fix?
you may as well remove this full, 40 defeats the purpose of it

File: src/main/java/uk/gov/hmcts/reform/idam/client/models/GeneratePinRequest.java [hunk 5]
```diff
@@ -11,11 +11,11 @@
 @Builder
 @AllArgsConstructor
 public class GeneratePinRequest {
-
     private final String firstName;
     @JsonInclude
     private final String lastName;
-    private final List<String> roles;
```
Comment (timja) [inline src/main/java/uk/gov/hmcts/reform/idam/client/models/GeneratePinRequest.java]: why make this not final?

Comment (alexwakeman) [reply src/main/java/uk/gov/hmcts/reform/idam/client/models/GeneratePinRequest.java]: Was left-over, have reverted

File: src/main/java/uk/gov/hmcts/reform/idam/client/IdamApi.java (hunk 1 above)
Comment (alexwakeman) [reply src/main/java/uk/gov/hmcts/reform/idam/client/IdamApi.java]: 👍

**Pull request merged**